### PR TITLE
MUU-42: Notifications to client

### PR DESCRIPTION
**Pull request by natlibfi-kmskuusi**
Purpose of this larger ticket set is to enable server side notifications to client , that will show user either dialog or banner to inform user on different matters, including ones blocking usage i.e. server outage. Multiple sub task tickets included within pr. Tickets will have link to this single pr.

Added to scope rework of the snackbar/banner ui logic.

### CODE REVIEW

File: src/clients/common/notificationBanner.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+
+let timeoutId = null;
+
+function createNotificationBanner(notificationBannerContent, html) {
+  const notificationBannerElement = getNewNotificationBanner(html);
+
+  const notificationBannerType = checkNotificationBannerType(notificationBannerContent);
+
+  switch (true) {
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationBanner.js:80]: avoid Switch, use if + return. If needed make helper function

File: src/clients/common/notificationBanner.js [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        iconColor = 'var(--color-blue-100)';
+        icon = 'warning_amber';
+      }
+
+      if (style === 'error') {
+        backgroundColor = 'var(--color-red-80)';
+        iconColor = 'var(--color-red-100)';
+        icon = 'report_gmailerrorred';
+      }
+
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationBanner.js:170]: create helper function that return object containing all 3 colors and takes style as param. If param is undefined return default style

File: src/clients/common/notificationBannerStatic.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+  }
+  function styleBasedOnType(){
+    
+    let backgroundColor = '';
+    let iconColor = '';
+    let icon = '';
+
+    switch (content.style) {
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationBannerStatic.js:62]: avoid switch and use style helper function mentioned in `src/clients/common/notificationBanner.js`

File: src/clients/common/notificationDialog.js [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+/**
+ * 
+ * @param {*} notificationDialogContent object
+ * @param {*} notificationDialogContent.id id of dataobject if any
+ * @param {*} notificationDialogContent.style info/success/alert/error
+ * @param {*} notificationDialogContent.text given text
+ * @param {*} isStatic should dialog fade out ?
+ * @param {*} canDismiss show close dialog button ?
+ * @param {*} showBackground show semi transparent backgroudn blocking layer
+ * @returns
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationDialog.js:45]: Param types

File: src/clients/common/notificationDialog.js [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+
+let timeoutId = null;
+
+function createNotificationDialog(notificationDialogContent, html, isStatic = true, canDismiss = false) {
+  const notificationDialogElement = getNewNotificationDialog(html);
+
+  const notificationDialogType = checkNotificationDialogType(notificationDialogContent);
+
+  switch (true) {
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationDialog.js:104]: avoid switch

File: src/clients/common/notificationDialog.js [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  function createCustomNotificationDialog(notificationDialogObject) {
+    const {style, text, linkButton, title} = notificationDialogObject;
+
+    setNotificationDialogStyle();
+    addTextToNotificationDialog(text);
+    addLinkButton();
+
+    //if title is provided use that but if not use default
+    if(title){notificationDialogElement.querySelector(`.notificationdialog-title`).innerHTML = titleText;}
+    else{
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationDialog.js:170]: avoid else.. use return

File: src/clients/common/notificationDialog.js [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        backgroundColor = 'var(--color-yellow-80)';
+        iconColor = 'var(--color-blue-100)';
+        icon = 'warning_amber';
+      }
+
+      if (style === 'error') {
+        backgroundColor = 'var(--color-red-80)';
+        iconColor = 'var(--color-red-100)';
+        icon = 'report_gmailerrorred';
+      }
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationDialog.js:206]: use style helper function mentioned in `src/clients/common/notificationBanner.js`

File: src/clients/common/notificationDialog.js [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+    function addLinkButton() {
+      if (linkButton !== undefined && linkButton.nodeName === 'BUTTON') {
+        notificationDialogElement.querySelector(`.notificationdialog-link`).style.display = 'flex';
+        notificationDialogElement.querySelector(`.notificationdialog-link`).append(linkButton);
+      }
+    }
+
+    function addDefaultStyleBasedTitle(){
+      let titleText = '';
+      switch (style) {
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationDialog.js:223]: avoid switch

File: src/clients/common/notificationDialog.js [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+      //in case there are blocking dialogs that have userinteraction enabled
+      //check if there is any visible dialogs after closing one return the background normal
+      if(!hasActiveChildren()){
+        hideBackground();
+      }
+      
+      function hasActiveChildren(){
+        const notificationDialogsContainer = document.getElementById('notificationDialogs');
+        const children = notificationDialogsContainer.children;
+        for (const child of children) {
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationDialog.js:266]: children.forEach(child => { ...

File: src/clients/common/notificationDialog.js [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  function displayNotificationDialog(isStatic = true) {
+    const notificationDialog = document.querySelector('#notificationDialog');
+
+    listenToNotificationDialogAnimationStart();
+    listenToNotificationDialogAnimationEnd();
+    if(isStatic){
+        notificationDialog.classList.add('show');
+    }else{
```
Comment (natlibfi-jonollil) [inline src/clients/common/notificationDialog.js:292]: avoid else

**Review by natlibfi-jonollil — CHANGES_REQUESTED** [review]

File: src/clients/common/notificationBanner.js (hunk 1 above)
Comment (natlibfi-kmskuusi) [reply src/clients/common/notificationBanner.js]: Updated. Switching away from switch

File: src/clients/common/notificationBanner.js (hunk 2 above)
Comment (natlibfi-kmskuusi) [reply src/clients/common/notificationBanner.js:179]: Now using helper function and using returned object in banner/banner_static and dialog.

File: src/clients/common/notificationDialog.js (hunk 4 above)
Comment (natlibfi-kmskuusi) [reply src/clients/common/notificationDialog.js:45]: Missing param types updated

File: src/clients/common/notificationDialog.js (hunk 6 above)
Comment (natlibfi-kmskuusi) [reply src/clients/common/notificationDialog.js]: Using helper functions with return statements now avoiding else

**Comment by natlibfi-kmskuusi** [comment]
Screenshots of common use case ui elements included to MUU-348

**Comment by natlibfi-kmskuusi** [comment]
NOTE!
For notifications route to be able to properly function enviroment variable "NOTE_MONGO_URI" should be set. 

If not set routes mongo operator will result to undefined and usage in "get" catches it and defaults to returning empty array.

**Review by natlibfi-koskimim — APPROVED** [review]

**Pull request merged**